metadata/src/main/java/org/apache/kafka/controller/QuorumController.java

```diff
@@ -301,6 +328,89 @@ private void appendControlEvent(String name, Runnable handler) {
         queue.append(event);
     }
 
+    private static final String GENERATE_SNAPSHOT = "generateSnapshot";
+
+    private static final int MAX_BATCHES_PER_GENERATE_CALL = 10;
+
+    class SnapshotGeneratorManager implements Runnable {
+        private final Function<Long, SnapshotWriter> writerBuilder;
+        private final ExponentialBackoff exponentialBackoff =
+            new ExponentialBackoff(10, 2, 5000, 0);
+        private SnapshotGenerator generator = null;
+
+        SnapshotGeneratorManager(Function<Long, SnapshotWriter> writerBuilder) {
+            this.writerBuilder = writerBuilder;
+        }
+
+        void createSnapshotGenerator(long epoch) {
```

Comment:
Do we need to check to see if this epoch is known or valid?

Reply:
It's not technically necessary since we have other checks, but I will add a check.  In general this situation should not occur.